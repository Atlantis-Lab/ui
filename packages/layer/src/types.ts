import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  FlexboxProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
} from 'styled-system'

export interface LayerProps
  extends BorderProps,
    ColorProps,
    LayoutProps,
    ShadowProps,
    PositionProps,
    BackgroundProps,
    SpaceProps,
    FlexboxProps {
  appRoot?: string
  active: boolean
  setActive: (payload: any) => any
  outsideClick?: boolean
  children: any
}
