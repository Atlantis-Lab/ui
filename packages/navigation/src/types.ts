export interface ContainerProps {
    width?: string
    height?: string
}

export interface ControlsProps {
    amount: number
    width?: string
    height?: string
    spaceBetween?: number
    activeFragment?: number
    backgroundColor?: string
    activeBackgroundColor?: string
}
