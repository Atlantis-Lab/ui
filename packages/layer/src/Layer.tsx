import React, { useEffect }                                                    from 'react'
import { createPortal }                                                        from 'react-dom'
import { background, border, color, flexbox, layout, position, shadow, space } from 'styled-system'

import styled                                                                  from '@emotion/styled'

import { LayerProps }                                                          from './types'

const Wrapper = styled.div(border, layout, space, flexbox, color, shadow, position, background)

const Layer = ({ appRoot, active, setActive, children, outsideClick, ...props }: LayerProps) => {
  const container = document.createElement('div')
  const root = document.getElementById(appRoot)

  const unmountLayer = () => {
    if (root.contains(container)) {
      root.removeChild(container)
      window.removeEventListener('click', unmountLayer)
      setActive(false)
    }
  }

  const onOutsideClick = outsideClick
    ? (event) => !container.contains(event.target) && unmountLayer()
    : () => {}

  useEffect(() => {
    if (active) {
      window.addEventListener('click', onOutsideClick)
      root.appendChild(container)
    }
    return () => {
      unmountLayer()
      window.removeEventListener('click', onOutsideClick)
    }
  }, [active])

  return createPortal(<Wrapper {...props}>{children}</Wrapper>, container)
}

Layer.defaultProps = {
  appRoot: 'root',
  active: false,
  outsideClick: true,
}

export default Layer
