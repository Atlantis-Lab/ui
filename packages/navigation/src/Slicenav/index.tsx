import React, { FC }     from 'react'

import { Box }           from '@atlantis-lab/layout'

import { ControlsProps } from "../types"
import { Container }     from './Container'
import { Slice }         from './Slice'

export const SliceNav: FC<ControlsProps> = ({
    amount,
    width = '100%',
    height = '100%',
    spaceBetween = 5,
    activeFragment,
    backgroundColor = 'grey',
    activeBackgroundColor = 'green',
    wereActiveBackgroundColor = 'green',
    keepActive = false,
                                          }) => {
    let elements = []

    for (let i = 0; i < amount; i+=1) {
        elements.push(i)
    }

    const purgedWidth = Number(width.replace(/(%|px|vh|rem|em)/,''))

    const sliceWidth = purgedWidth/amount-spaceBetween

    return (
    <Container width={width} height={height}>
        {
            elements.map(element => <Slice
                key={element}
                width={sliceWidth}
                backgroundColor={
                    keepActive
                    ?
                    activeFragment-1>=element ? wereActiveBackgroundColor : backgroundColor
                    :
                    activeFragment-1===element ? activeBackgroundColor : backgroundColor
                }
            />)
        }
    </Container>
        )
}
