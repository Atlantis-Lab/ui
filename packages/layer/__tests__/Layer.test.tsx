import React     from 'react'
import ReactDOM  from 'react-dom'
import renderer  from 'react-test-renderer'

import { Layer } from '../src'

describe('Layer', () => {
  it('should render layer correctly', () => {
    ReactDOM.createPortal = (node) => node

    const tree = renderer
      .create(
        <div id='appRoot'>
          <Layer setActive={() => {}} appRoot='appRoot' border='1px solid black'>
            Portal
          </Layer>
        </div>,
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
