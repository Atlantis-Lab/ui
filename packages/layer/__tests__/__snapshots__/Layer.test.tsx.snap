// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer should render layer correctly 1`] = `
.emotion-0 {
  border: 1px solid black;
}

<div
  id="appRoot"
>
  <div
    className="emotion-0"
  >
    Portal
  </div>
</div>
`;
